### ISSUE TITLE
How to safely interrupt model training (e.g. for Speaker Encoder, using the LibriSpeech dataset)? Is loss below 0.05 considered good enough?  

### ISSUE THREAD

**Issue by goodmangu** [CorentinJ/Real-Time-Voice-Cloning#379]
I have been training the Speaker Encoder model using the LibriSpeech dataset for nearly a whole day, for well over 200K steps. Loss is below 0.03 and EER is well below 0.005. Is this considered good enough? If not, what loss level is considered excellent? How to safely terminate training?

**Comment by blue-fish** [comment]
Take a look at #126 if you haven't already, there is some discussion about training the speaker encoder.

**Comment by notld** [comment]
Can anyone share from experience how many steps are involved in just the voxceleb1 concatenated dataset for the encoder? I’m on step 4080 and counting and it’s been a couple days. Along the way during one the sets I received text from PowerShell that I wasn’t using Intel’s Threading Building Blocks version 2019.5 or higher (~I believe im using 1101), and it just continued.

**Comment by goodmangu** [comment]
@CorentinJ : in the Jupyter notebook, it is mentioned that encoder "pretrained.pt" was trained to step 1,564,501. Could you advise on that the loss was at that stage?

**Comment by blue-fish** [comment]
You can resume training of the model to determine the loss.

**Comment by goodmangu** [comment]
How do you safely pause/terminate model training?

**Comment by blue-fish** [comment]
If you are using the defaults, per [encoder_train.py](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/070a3c187f87136ebe92aa72766f8343772d414e/encoder_train.py#L28) it saves to pretrained.pt every 500 steps (and a backup to a separate checkpoint file every 7,500 steps). As long as you don't pass the --force_restart flag, it will reload pretrained.pt and continue training where it left off.

**Comment by goodmangu** [comment]
cool. Thanks @blue-fish! What is the best loss you have got so far, and at what steps?

**Comment by blue-fish** [comment]
I have not yet trained a speaker encoder, so I don't have any experience to share. Still new to this. Currently trying to find out if the pretrained vocoder improves with additional training.

**Comment by CorentinJ** [comment]
Check the plots for the speaker encoder. Are the clusters well separated in the space? If so, your model is good enough

**Comment by goodmangu** [comment]
Thanks @CorentinJ : This is the UMAP Projection plot persisted and it appears the clusters are well separated in space?
![my_run_umap_219900](https://user-images.githubusercontent.com/11790331/85903767-2848d400-b7d5-11ea-9fb7-76c7d3c659ef.png)

**Comment by CorentinJ** [comment]
Yes, you can stop the training there.

**Comment by goodmangu** [comment]
Thanks @CorentinJ . Do you have UMAP plot for the best pre-trained model that you have shared and loss at that point?

**Comment by goodmangu** [comment]
These are the last two I got:
Encoding model training 
..........
Step 219910   Loss: 0.0274   EER: 0.0046   Step time:  mean:   658ms  std:   632ms

Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:  125ms   std:   13ms
  Data to cuda (10/10):                            mean:    3ms   std:    0ms
  Forward pass (10/10):                            mean:   38ms   std:    0ms
  Loss (10/10):                                    mean:   22ms   std:    1ms
  Backward pass (10/10):                           mean:  134ms   std:   22ms
  Parameter update (10/10):                        mean:  124ms   std:    3ms
  Extras (visualizations, saving) (10/10):         mean:   42ms   std:  126ms

..........
Step 219920   Loss: 0.0294   EER: 0.0050   Step time:  mean:   639ms  std:   472ms

Average execution time over 10 steps:
  Blocking, waiting for batch (threaded) (10/10):  mean:  284ms   std:  475ms
  Data to cuda (10/10):                            mean:    3ms   std:    0ms
  Forward pass (10/10):                            mean:   38ms   std:    0ms
  Loss (10/10):                                    mean:   22ms   std:    2ms
  Backward pass (10/10):                           mean:  126ms   std:   20ms
  Parameter update (10/10):                        mean:  123ms   std:    5ms
  Extras (visualizations, saving) (10/10):         mean:   36ms   std:  108ms

**Comment by CorentinJ** [comment]
Yeah you'll find some in my thesis

**Comment by blue-fish** [comment]
Closing this issue as it appears to be resolved. @goodmangu Please reopen this issue if you continue to have additional questions.
